Merge wip with master

tehw0lf opened the pull request:
The wip branch containing v1.1 should be ready to be merged into master.
This update will bring library functionality and can be installed via pip.

tehw0lf committed:
Move unit tests

Relocated unit tests and associated files to ./test/ to optimize project structure

Moving towards #3

tehw0lf committed:
Class is now instantiated

As described in issue #5, the class was never instantiated, which was resolved by this commit.

Fixes #5

tehw0lf committed:
Move sample file

Relocated sample.txt to ./res/sample.txt to keep the root directory clean

Moving towards #3

tehw0lf committed:
Removed self.report

The variable self.report was only used as container for the amount of warnings stored in self.wcount, so it could be removed by simply printing out self.wcount directly.

Fixes #7

tehw0lf committed:
Migrate parse_args() out of the class

The argument parsing function parse_args() was a part of Paw, which prevented library use.

Fixes #8 and partially resolves #6

Commit:
Update .gitignore

Replace obsolete .gitignore entry with .vscode

tehw0lf committed:
Implement library functionality

Prior to this commit, paw could only be used as standalone tool. Now it can be used as a library as well.

Fixes #6

tehw0lf committed:
Restructure project data

The code is no longer unorganized, but structured into subfolders that allow installation via setuptools

Resolves #3

tehw0lf committed:
Update .gitignore

Added test directory to blacklist.

tehw0lf committed:
Add setup.py

A setup file was added to allow installation via pip. Also updated travis.yml to reflect the changes.

Fixes #4

tehw0lf committed:
Move paw to paw/__init__.py

Now the code is restructured.

tehw0lf committed:
Fix .gitignore

By mistake, the whole test_files directory was added to .gitignore, which would cause (future) tests to fail.

tehw0lf committed:
Update .gitignore

Blacklisted the .egg-info/ directory created by setuptools.

tehw0lf committed:
Separate cli from library

Split code into main library file and commmand line interface file.

tehw0lf committed:
Fix command line script

The command line script is now usable.

tehw0lf committed:
Re-enable cli functionality

The setup process will now install paw as command line tool.

tehw0lf committed:
Fix crash possibility

In command line mode, the program would crash if invoked with either -c or -p without providing an input file.

Spindel commented on paw/__init__.py:
```diff
@@ -0,0 +1,193 @@
+            for i in self.wlist:
+                print(i)
+
+    def __init__(self, gensets=None, hcat=False, infile=None, outfile=None):
```
Convention is that dunder `__[a-z]+__` functions are declared first in the class

Spindel commented on paw/__init__.py:
```diff
@@ -0,0 +1,193 @@
+                self.catstrs[k] = (catstr + hexu + hexl +
+                                   pattern.replace('??', '?'))
+            else:
+                print('warning: pattern %d is empty' % k)
```
This should use logging module at the  `warning` level rather than `print`

Spindel commented on paw/__init__.py:
```diff
@@ -0,0 +1,193 @@
+        Generate pattern and position specific charset from string
+        '''
+        # Get pattern for each string position
+        pattern = ['%' + self.cset_lookup(i) for i in instr]
```
This is only used as an iterator, so can be changed to 
`pattern = ('%' + self.cset_lookup(i) for i in instr)`

Spindel commented on paw/__init__.py:
```diff
@@ -0,0 +1,193 @@
+        pattern = [''.join(sorted(set(i))) for i in pattern]
+        for i in range(len(instr)):
+            try:
+                self.cset[i] = ''.join(sorted(set(self.cset[i]
```
You're sprinkling the code with `sorted(set` in a lot of places.

tehw0lf replied:
Good point. I will look into it and see if I can just use sets while still preserving the position specific charsets.

Spindel commented on paw/__init__.py:
```diff
@@ -0,0 +1,193 @@
+
+        # Sort pattern and keep only unique values
+        pattern = [''.join(sorted(set(i))) for i in pattern]
+        for i in range(len(instr)):
```
Didn't you just once work through instr?   And now you're doing it again?

tehw0lf replied:
Indeed, though the first time was just to create the pattern, this one works on the charset that will be used, unless I am missing something

Spindel replied:
Yeah, the question is more, why is your algorithm walking over the same data repeatedly when it could walk over it once.

tehw0lf replied:
Oh well, that's something I missed. The first time is to create the pattern from the input string, the second time reads the values again to build the character set. I guess it would be better to move the pattern creation into the for loop and do both things at the same time.

Spindel commented on paw/__init__.py:
```diff
@@ -0,0 +1,193 @@
+                cnt -= 1
+                cur += 1
+            else:
+                if tmpsets[i] == '%':
```
This could be an elif to the above else.

Spindel commented on paw/__init__.py:
```diff
@@ -0,0 +1,193 @@
+                    except KeyError:
+                        self.cset[cur] = tmpsets[i]
+        if cnt > 0:
+            print('warning: input contains uneven number of brackets')
```
Should be logging not print

Spindel commented on paw/__init__.py:
```diff
@@ -0,0 +1,193 @@
+            print('warning: input contains uneven number of brackets')
+            self.wcount += 1
+
+    def save_wordlist(self):
```
Since this is the only place that uses outfile, why not make outfile an argument to this function?

Spindel reviewed (commented):
Good improvements, and a nice test suite too.

I'm a bit concerned about the many calls to set of things, wouldn't it be better to just use sets as the native data store then, rather than to cast and recast repeatedly?

tehw0lf committed:
Make outfile an argument to save_wordlist

Outfile is not needed anywhere else, so it is now specific to this function.

Part of #9

tehw0lf committed:
Move __init__ to the top

As per convention, the dunder function has been moved to the top of the class.

Part of #9

tehw0lf committed:
Use logging instead of print

The logging module provides an interface for warning messages that should be used instead of plain print()

part of #9

tehw0lf committed:
Use a generator expression

Instead of a list comprehension, pattern can be changed to be a generator expression without losing functionality

Part of #9

tehw0lf committed:
Remove duplicate init function

The function was copied to the top, yet not deleted at the bottom

Part of #9

tehw0lf committed:
Rework if/elif structure

Else: if: was not needed, elif could be used in this case.

Part of #9

tehw0lf committed:
Reduce calls to sorted(set())

The combination sorted(set()) was used in many places, albeit not being needed in all of them.

Part of #9

tehw0lf committed:
Remove outfile from initial call

Since outfile is now an argument to save_wordlist(), it has to be removed from the call to Paw()

tehw0lf committed:
Update License

The license was updated to include the current year.

tehw0lf committed:
Rename gen_wordlist() argument

arr2d was renamed to cset_dict to enhance readability.

Part of #9

tehw0lf committed:
Update readme

Now reflects new features and documents both command line and library use.

Part of #9

tehw0lf committed:
Update screenshot

Now the screenshot correctly shows the new command line interface instead of the older standalone version.

Part of #9

tehw0lf committed:
Fix coverage command

Faulty command replaced with automated test discovery.

Part of #9

tehw0lf committed:
Introduce line number report

Paw will now report the number of lines that will be written.

Part of #9

tehw0lf committed:
Fix the screenshot

The cropping was faulty.

Part of #9

tehw0lf committed:
Update .gitignore

Add build and dist dir to blacklist

Part of #9

tehw0lf committed:
Replace gen_wordlist()

A new algorithm to generate wordlists is introduced which generates one word at a time, thus ensuring that memory hogging does not happen.

Resolves #10

tehw0lf committed:
Fix gen_wordlist()

The previous function called save_word() on each generated word. This version can be iterated over as it now is a proper Python generator.

Fixes #10

tehw0lf committed:
Add default value to outfile

The output file name should be None by default.

tehw0lf committed:
Fix charset generation

All functions now work properly with the new word generating algorithm.

tehw0lf committed:
Update unit tests

The unit tests were adapted to reflect the new changes.

Fixes #12

tehw0lf committed:
Update readme

The documentation was updated.

Commit:
Update readme

Fixed the variable name and wording.

tehw0lf committed:
Update travis configuration

To run the updated tests, mock needs to be installed.

tehw0lf committed:
Optimize code

Instead of range and len, enumerate is now used for iteration. Also the position array is using integer values instead of strings now.

Part of #9

tehw0lf committed:
Replace line calculation method

The incrementing counter was replaced by a simple calculation based on the input character sets,

Resolves #13

tehw0lf committed:
Build word by list comp

Using join() and list comprehension performs significantly better than iterating over the list and appending to a string.

Part of #14

tehw0lf committed:
Move generator out of class

The wordlist generator is now standalone, as the implementation does not require it to be a part of the class.

Part of #14

tehw0lf committed:
Remove functions from class

The generator was removed from the class, and the tests were adjusted to reflect the changes.

Part of #14

tehw0lf committed:
Remove deprecated variable

The old line counter was still used in command_line.py, causing it to fail with an error despite correctly saving the wordlist.

tehw0lf committed:
Remove wgen

Instead of being a part of paw, wgen now has its own repo at tehw0lf/wlgen

tehw0lf committed:
Fix imports and change default method

wlgen provides two recursive implementations and one using itertools.product. For now, the latter is set as default, because it performs good with low memory usage.

tehw0lf committed:
Add wordlist generator to __init__

The generator function that should be used can now be defined in __init__.py. In the future, this will become a command line switch.

tehw0lf committed:
Add dependencies

The automated download of wlgen was not implemented yet.

tehw0lf committed:
Fix formatting

setup.py now passes flake8 again

tehw0lf committed:
Fix typo

setup.py now works again

tehw0lf committed:
Introduce buffering

After implementing a buffer list, a major speedup could be observed when using a buffer size of 256.

Resolves #14, part of #9

tehw0lf committed:
Introduce buffering

After implementing a buffer list, a major speedup could be observed when using a buffer size of 256.

Resolves #14, part of #9

tehw0lf committed:
Merge branch 'wip' of https://github.com/tehw0lf/paw into wip

tehw0lf committed:
Update .gitignore

tehw0lf committed:
Fix CLI

Removed the duplicate -b argument from the command line interface.

Closes #9

tehw0lf committed:
Update README.md

tehw0lf committed:
Add algorithm parameter to CLI

The algorithm that will be used for generating the wordlist can now be chosen by using the -a parameter.

tehw0lf committed:
Merge branch 'wip' of https://github.com/tehw0lf/paw into wip

tehw0lf committed:
Update screenshot

tehw0lf committed:
Update unit tests

The unit tests were updated to test all algorithms that are imported from wlgen.
